Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h

```diff
@@ -210,6 +210,37 @@ static inline void emitCatchPrologueShared(B3::Air::Code& code, CCallHelpers& ji
     jit.addPtr(CCallHelpers::TrustedImm32(-code.frameSize()), GPRInfo::callFrameRegister, CCallHelpers::stackPointerRegister);
 }
 
+static inline void prepareForReturnCall(CCallHelpers& jit, const B3::StackmapGenerationParams& params, const Checked<int32_t>& newStackOffset)
+{
+    GPRReg scratch = jit.scratchRegister();
+    Checked<int32_t> frameSize = params.code().frameSize();
+    Checked<int32_t> stackOffset = frameSize + newStackOffset;
+
+    jit.emitRestore(params.code().calleeSaveRegisterAtOffsetList());
+
+#if CPU(X86_64)
+    jit.move(CCallHelpers::Address(MacroAssembler::framePointerRegister, CallFrame::returnPCOffset()),
+        CCallHelpers::Address(MacroAssembler::framePointerRegister, newStackOffset - CallFrame::returnPCOffset()), scratch);
+    stackOffset -= Checked<int32_t>(CallFrame::returnPCOffset());
+#elif CPU(ARM64) || CPU(APPLE_ARMV7S) || CPU(RISCV64)
+    jit.loadPtr(CCallHelpers::Address(MacroAssembler::framePointerRegister, CallFrame::returnPCOffset()), MacroAssembler::linkRegister);
+#if CPU(ARM64E)
+    jit.addPtr(MacroAssembler::TrustedImm32(sizeof(CallerFrameAndPC)), MacroAssembler::framePointerRegister, scratch);
+    jit.untagPtr(scratch, MacroAssembler::linkRegister);
+    jit.validateUntaggedPtr(MacroAssembler::linkRegister, scratch);
+#endif
+#else
+    UNREACHABLE_FOR_PLATFORM();
+#endif
+    jit.loadPtr(CCallHelpers::Address(MacroAssembler::framePointerRegister), MacroAssembler::framePointerRegister);
+    for (unsigned i = 0; i < params.size(); ++i) {
+        if (params[i].isStack()) {
+            jit.move(CCallHelpers::Address(MacroAssembler::stackPointerRegister, frameSize + params[i].offsetFromFP()),
```

Comment:
Can we avoid this by emitting a stack overflow check and bumping the stack pointer sooner, then shuffle the arguments into place?

Reply:
Also related to comment on `createReturnCallPatchpoint` I think.

Reply:
Actually, I _think_ we can deal with all arguments in place sooner, by passing the exceeding arguments as locked stack slots to the patchpoint. It seems to work with return_call.